### PR TITLE
refactor backend and use NestJs

### LINKED ISSUE #43: Setup NestJs for the backend
I've been keeping an eye on [NestJs](https://github.com/nestjs/nest) for few months and it seems to be really amazing.

Typescript, decorators, on top of express, websocket integration, syntax similar to Angular... Pizza-Sync seems to be the perfect candidate for a first app with NestJs!

### PR DESCRIPTION
**Pull request by maxime1992**
This refactor has gone wild :sweat_smile:.

I wanted to have a clean git history but at some point I just loose the track and ended up with WIP that I'll fixup at the end.

That said, here's what's this PR will bring to Pizza-Sync:
- Full refactor of the backend with [NestJs](https://nestjs.com)
- Way better server architecture _*1_
- Possibility to quickly add new commands
- Possibility to add a local pizza provider (mock) for testing or offline developing :+1: 

**TODO**

- [x] ingredients appear multiple time in frontend (--> they're not well parsed on the backend)
- [x] ingredients are not ordered by their names
- [ ] separate the backend in two folders: core and features?
_this would be helpful for people who want to contribute either to the core (adding functionnalities) or features (pizzas providers, comands)_
- [ ] add some tests
- [x] it's possible to change a provider when someone is already connected and even ordered
- [x] `backend/conf.json` has not been reimplemented

_*1_  At the beginning, backend was not something I wanted to showcase on Pizza-Sync and I made a really quick implementation (pretty dirty). Few months later, I heard of NestJs, the Typescript starter and the syntax very similar to Angular and it seems that Pizza-Sync was the perfect opportunity to learn how to use Nest :tada: 

Closes #43

### CODE REVIEW

**Comment by maxime1992** [comment]
@albinscode @victornoel @tbille @strokyl anybody motivated for a review :sweat_smile:?

It's pretty huge and you should definitely clone the repo and checkout `back/nestjs` if you want to review.

I just deleted the backend folder and created a brand new one using Nest framework :smile:!

It's not finished yet (see PR description) but it's pretty stable for a review.

No problem if you're not interested or don't have time :+1: :)

**Comment by victornoel** [comment]
command is written with two m, not one :) (see directory `comand-line`)

File: backend/package.json [hunk 1]
```diff
@@ -1,26 +1,63 @@
 {
-  "name": "pizza-sync",
+  "name": "nest-typescript-starter",
   "version": "1.0.0",
```
Comment (tbille) [inline backend/package.json]: this is a big change no version update?
(Really really useful comment 😄 )

File: backend/src/features/models/users/users.component.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+      return { event: 'CONNECT_USER_SUCCESS', data: user };
+    } else {
+      const newUser = await this.addUser(username);
+
+      client.user = newUser;
+
+      this.setUserOnline(newUser);
+      return { event: 'CONNECT_USER_SUCCESS', data: newUser };
+      // this.server.sockets.emit('CONNECT_USER_SUCCESS', newUser);
```
Comment (tbille) [inline backend/src/features/models/users/users.component.ts]: I guess you can remove this line

File: backend/src/features/models/users/users.component.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+  getUser(username: string): IUserWithId {
+    const user = this.allIds
+      .map(userId => this.byId[userId])
+      .find(userTmp => userTmp.username === username);
+
+    if (user) {
+      return user;
+    }
+
+    return null;
```
Comment (tbille) [inline backend/src/features/models/users/users.component.ts]: `return user ? user : null;`

**Comment by tbille** [comment]
I added a few simple comments :) 
This is a big PR, but you should add a few steps to help developers to QA you PR. For example here, tell us how to test the new system of events you have in the frontend.

And woaw, I did a quick review but nest looks great!

File: backend/package.json (hunk 1 above)
Comment (albinscode) [reply backend/package.json]: Yeah, it desserves at least a 2.0.0 version ! :)

File: backend/src/command-line/command-line.component.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      new ProvidersCommand(this.vorpal, pizzasProvidersService),
+      new CountdownCommand(this.vorpal, ordersService)
+    );
+
+    // register every commands
+    this.commands.forEach(command => command.register());
+
+    this.vorpal.delimiter('pizza-sync$').show();
+  }
+}
```
Comment (albinscode) [inline backend/src/command-line/command-line.component.ts]: Well it is a better design to group commands. Injection part is very good too thanks to nestjs !

**Comment by maxime1992** [comment]
@albinscode yes I'm really interested in bringing some tests in the backend. That's definitely one of the good thing I was expecting with Nest!

I'll try to clean up a little bit and take @tbille comments into account.
If I feel like it's not too bad, I'll just merge the PR and we could add tests and even improvements in further PR instead of this one, what do you think?

File: backend/src/features/models/users/users.component.ts (hunk 2 above)
Comment (maxime1992) [reply backend/src/features/models/users/users.component.ts]: Done

File: backend/src/features/models/users/users.component.ts (hunk 3 above)
Comment (maxime1992) [reply backend/src/features/models/users/users.component.ts]: Done

**Comment by maxime1992** [comment]
Thanks all for reviewing :smile:, I've just merged this branch and released v2.0.0 (see [changelog](https://github.com/maxime1992/pizza-sync/blob/master/CHANGELOG.md#v200-2017-12-27))

File: backend/package.json (hunk 1 above)
Comment (maxime1992) [reply backend/package.json]: Bumped to 2.0.0! thx

**Comment by albinscode** [comment]
Good news! Next, I will submit you my new PR about another pizza provider and then I'll try adding some tests, thanks!

**Pull request merged**